lupa/_lupa.pyx

```diff
@@ -1725,6 +1817,44 @@ cdef tuple unpack_multiple_lua_results(LuaRuntime runtime, lua_State *L, int nar
     return args
 
 
+# bounded memory allocation
+
+cdef void* _lua_alloc_restricted(void* ud, void* ptr, size_t old_size, size_t new_size) nogil:
+    # adapted from https://stackoverflow.com/a/9672205
+    cdef size_t* memory_left = <size_t*>ud
+
+    if ptr is NULL:
+        # <http://www.lua.org/manual/5.2/manual.html#lua_Alloc>:
+        # When ptr is NULL, old_size encodes the kind of object that Lua is allocating.
+        # Since we don’t care about that, just mark it as 0.
+        old_size = 0
+
+    if new_size == 0:
+        free(ptr)
+        if memory_left[0] > 0:
+            memory_left[0] += old_size  # add old size to available memory
+        return NULL
+    elif new_size == old_size:
+        return ptr
+    else:
+        if memory_left[0] > 0 and new_size > old_size and memory_left[0] <= new_size - old_size:  # reached the limit
+            return NULL
+        new_ptr = realloc(ptr, new_size)
+        if new_ptr is not NULL and memory_left[0] > 0:
+            memory_left[0] += new_size - old_size
+        return new_ptr
```

Comment:
I changed two things here.

1) `realloc()` does not `free()` the memory on failures, and I think we shouldn't do that here either. According to the [example in the Lua docs](https://www.lua.org/manual/5.4/manual.html#lua_Alloc), Lua uses the plain `realloc()` behaviour. Therefore, I removed the `free()`. I would assume Lua to call into our allocator again in order to free the old pointer on failures, if it decides to free it.

2) The original calculation was `memory_left[0] -= new_size + old_size`, which looked wrong to me. It should add the difference, not subtract the sum.

Reply:
You are correct - the logic is wrong, but so is yours.
It should be `memory_left[0] -= new_size - old_size`.

```py
# allocate 20 more bytes
memory_left, new, old = 100, 50, 30
print(memory_left - (new - old))  # 80 (correct)
print(memory_left - (new + old))  # 20
print(memory_left + (new - old))  # 120
print(memory_left + (new + old))  # 180
```

Reply:
Right. It says "memory left", not "memory used". I mixed that up.

Reply:
While cleaning it up a bit, I noticed that we stop counting the memory when users disable it, but then, when users enable it again, we start from whatever "memory left" status we had when it was disabled. That doesn't sound clean. The current "memory left" might be arbitrarily out of sync with the actual memory usage at that point. Especially when you think of future memory deallocations of memory that was allocated during the time without memory limits.

Why not *always* count the memory that is *used* instead of (most of the time) the memory that is *left*, and then compare it to whatever the maximum allows memory usage is at the point of the allocation?

Reply:
Or, assume that the current memory usage is always `0` when users set a memory limit, and just take care of the difference from that point on.

Reply:
Because it was easier - `ud` can only point to one thing and I chose an size_t for that in the beginning when things were simpler.
We could probably change it to point to the LuaRuntime struct?

Reply:
> We could probably change it to point to the LuaRuntime struct?

That won't work because the allocator needs to run outside of the GIL. But we could use our own "memory allocation status struct", add such a struct value as a `LuaRuntime` attribute, and store the pointer to that struct in `ud`.

Reply:
We should then also include the initialisation phase in the memory counting, BTW, so that we cannot run into the case any more that anything gets deallocated that we didn't know about.